src/ProjectIdProvider.php

```diff
@@ -0,0 +1,206 @@
+    }
+
+    /**
+     * Check for a project id in the Key File specified in the GOOGLE_APPLICATION_CREDENTIALS
+     * environment variable
+     *
+     * @return string|null
+     */
+    private function fromKeyFilePath()
+    {
+        $credentials = CredentialsLoader::fromEnv();
+
+        if (!isset($credentials['project_id'])) {
+            return null;
+        }
+
+        return $credentials['project_id'];
+    }
+
+    /**
+     * Check for a project id in the application default credentials
+     *
+     * @return string|null
+     */
+    private function fromApplicationDefaultCredentials()
+    {
+        $credentials = CredentialsLoader::fromWellKnownFile();
+
+        if (!isset($credentials['project_id'])) {
+            return null;
+        }
+
+        return $credentials['project_id'];
+    }
+
+    /**
+     * Attempts to retrieve a project id from the gcloud sdk
+     *
+     * @return string|null
+     */
+    private function fromSdk()
+    {
+        $command = 'gcloud config config-helper --format json';
+
+        // exec the command
+        $output = $this->execute($command);
+
+        // if we didn't receive output, return null
+        if (!$output) {
+            return null;
+        }
+
+        // the output should be json. decode it and check for errors
+        $config = json_decode($output, true);
+
+        if (json_last_error() != JSON_ERROR_NONE) {
+            return null;
+        }
+
+        // the configuration should contain a project id at the specified path
+        if (!isset($config['configuration']['properties']['core']['project'])) {
+            return null;
+        }
+
+        return $config['configuration']['properties']['core']['project'];
+    }
+
+    /**
+     * Check for the project id available in App Engine Standard Environments
+     *
+     * @return string|null
+     */
+    private function fromAppEngineStandard()
+    {
+        // app engine should contain the app id environment variable
+        $appId = getenv('APPLICATION_ID');
+
+        // if we didn't find it, return null
+        if (!$appId) {
+            return null;
+        }
+
+        // find the project id embedded in the string
+        // looks something like p~project-name
+        $position = strpos($appId, '~');
+
+        if ($position > 0) {
+            $appId = substr($appId, $position + 1);
+        }
+
+        return $appId;
+    }
+
+    /**
+     * @return string|null
+     */
+    private function fromComputeEngineMetaData()
+    {
+        // Use the injected client if provided, or else instantiate one
+        $httpHandler = $this->httpHandler;
+
+        if (!$httpHandler) {
+            $httpHandler = HttpHandlerFactory::build();
+        }
+
+        // Make a request to the meta data service to get project id
+        // If we can't talk to the service and get a request exception, return null
+        try {
+            $response = $httpHandler(
```

Comment:
I think we'll need to follow the same pattern used in `GCECredentials` for talking to the metadata server, see [here](https://github.com/googleapis/google-auth-library-php/blob/master/src/Credentials/GCECredentials.php#L72-L83) for a little more detail.

Reply:
For what it's worth, GCECredentials doesn't apply those request settings to the real metadata calls. The retry count and timeout is used for the preflight request to check whether the application is on compute engine.

I think that rather than adding the timeout, this PR should call `GCECredentials::onGce()` prior to making the call to get the project ID. That way, future tweaks to make that method more reliable will cascade down to the method without us needing to keep things in sync.

Reply:
> this PR should call GCECredentials::onGce() prior to making the call to get the project ID.

👍

Reply:
Applied this change.